### ISSUE TITLE
PISM crashes with -subgl -bed_def iso,lc

### ISSUE THREAD

**Issue by tkleiner** [pism/pism#349]
PISM crashes with -subgl and -bed_def iso or -bed_def lc with error messages like,

`PISM ERROR: first  (X) equation in the SSAFD system: zero diagonal entry at a regular (not Dirichlet B.C.) location: i = XX, j = YY`

I am using stable v0.7.3-5-gf54025b (basic evolution run mode) on a cray cluster with PISM and PETSc compiled with Intel compliers and Intel-MPI. I did some tests with different geometries and found, that it is very difficult to reproduce the error in different runs. I finally used the circular ice sheet geometry from `$PISM_SRC_PATH=pism0.7/examples/marine/circular/circular_ice_sheet.py` without any additional option as input. Parameters for the PISM command are taken from the searise-antarctica example. 

This is the command:

``` bash
name=${SLURM_JOBID}_bed_lc
$PISM_MPIDO 4 pismr -i circular_withshelf.nc -bootstrap -mx 61 -my 61 -lz 5500 -lbz 2000 -mz 41 -mbz 11     -sia_e 3.0 -atmosphere given -surface given -ocean pik -meltfactor_pik 5e-3 -bed_def lc -ssa_method fd -ssa_e 0.6 -pik -calving eigen_calving,thickness_calving -eigen_calving_k 2.0e18 -thickness_calving_threshold 200.0 -stress_balance ssa+sia -pseudo_plastic -pseudo_plastic_q 0.25 -till_effective_fraction_overburden 0.02 -tauc_slippery_grounding_lines -topg_to_phi 15.0,40.0,-300.0,700.0 -subgl -hydrology null -age -ys -325000 -ye 0 -ts_file ${name}_ts.nc -ts_times -325000:1:0 -extra_file ${name}_ex.nc -extra_times -325000:1000:0 -extra_vars thk,usurf,mask,diffusivity,tauc,bmelt,tillwat,temppabase,hardav,tempicethk_basal,csurf,cbase,topg,Href -o ${name}.nc -o_size big
```

and [bed_lc_err.txt](https://github.com/pism/pism/files/349531/bed_lc_err.txt) is what I got after several runs with the **same command**. PISM crashes at different time steps and the error message is reporting different locations for different runs. 
Using `-bed_def iso`instead, PISM crashes again, but more reproducible (see [bed_iso_err.txt](https://github.com/pism/pism/files/349556/bed_iso_err.txt)). If I run PISM in two steps to avoid the critical year -0303851.445 (e.g. first from -325000 to -304000 and then restarting for -304000 to 0), PISM has no problems to pass -0303851.445. 

On my Mac with PISM and PETSc compiled with the GNU compiler, the runs (`-bed_def lc`) crash at very different time steps. Some of them are running through the entire time period of 325 kyrs. Ice thickness and ice extent are extremely varying over time in these runs.

Running PISM 0.7.3 (and pism-dev) in debug mode results in

```
S -0324965.487:   11.46658    5.12255          9.24260   4430.88166
PISM ERROR: ### Rank 0 message:
PISM ERROR: E=452784.797052 and pressure=6893446.836687 correspond to liquid water
PISM ERROR: ### Rank 0 message ends here.
PISM ERROR: Failure in a parallel section. See error messages above for more.
```

The exception is thrown from water_fraction_impl (enthalpyConverter.cc) but I have no idea, if this is causing the problem.

I have further checked out PISM stable0.6 (0.6.3) and found, that -subgl and -bed_def iso or -bed_def lc can be used without any problems. 

Thomas

**Comment by talbrecht** [comment]
I ran some test runs for Antarctica in stable0.7 (v0.7.3-2-gf64551e) on the PIK cluster with constant present-day climate forcing and had similar problems (if not using -calving ocean_kill -ocean_kill_file $initfile). I looked at those points where the SSA convergence failed and it seems that it occurs in grounded (island) points at the edges of the continental shelf (mask=2) almost surrounded by ic_free_ocean (7x mask=4), but where one of the four neighbors in a corner (e.g. i+1,j+1) becomes afloat (mask=2->3). 

At least, this pattern could give us a hint, where to look for the bug. Thomas, did you identify those failed points at the ice-ocean boundary, too? In fact, this case would cause a pretty tough entry in the SSA matrix (SSAFD.cc::assemble_matrix), where for the calving front boundary is accounted for (is_marginal()). I remember that we once introduced the iceberg remover, such that isolated (floating?) cells would not cause the SSA calculation to fail... -subgl helps to delay the timepoint of failure, but it cannot stop it. See http://www.pik-potsdam.de/~albrecht/bug07/

Torsten

**Comment by tkleiner** [comment]
Yes, the points are at the ice-ocean boundary. The searise-antarctica example is also broken on my machine. This example is without bed deformation but with -subgl (-ocean pik) enabled. 

```
cd $PISM_SRC_PATH/pism0.7/examples/searise-antarctica`
./preprocess.sh
./antspin-coarse.sh NN
```

I have tried `NN=36` and `NN=20`several times and found, that PISM fails in the run stage.
When the error occurs depends on the number of processors (consistent in repeated runs).

NN=36:
```
y  SSA:     1 outer iterations, ~25.0 KSP iterations each
v$Eh  diffusivity (overrides 2D CFL) (dt=1.19096)
S 00018024.584:   25.50953   13.47030          2.87555  13165.83881
PISM ERROR: ### Rank 22 message:
PISM ERROR: first  (X) equation in the SSAFD system: zero diagonal entry at a regular (not Dirichlet B.C.) location: i = 131, j = 155
            
PISM ERROR: ### Rank 22 message ends here.
PISM ERROR: stress balance computation failed. Model state was saved to 'run_30km_stressbalance_failed.nc'

#(antspin-coarse.sh)  coarse-grid part of spinup done
```

NN=20:
```
y  SSA:     3 outer iterations, ~22.0 KSP iterations each
v$Eh  2D CFL (dt=0.39903)
S 00000242.821:   25.51823   13.62510          4.15058  75124.22565
PISM ERROR: ### Rank 14 message:
PISM ERROR: first  (X) equation in the SSAFD system: zero diagonal entry at a regular (not Dirichlet B.C.) location: i = 114, j = 164
            
PISM ERROR: ### Rank 14 message ends here.
PISM ERROR: stress balance computation failed. Model state was saved to 'run_30km_stressbalance_failed.nc'

#(antspin-coarse.sh)  coarse-grid part of spinup done
```
In summery, bed deformation is not required to let PISM crash with -subgl in this version and I can't see the difference to PISM stable0.6.

Thomas

**Comment by talbrecht** [comment]
Right, the calving front boundary conditions (SSA-CFBC) ask for neighbors and if all four direct neighbors are ice free, the diagonal entry in the matrix vanishes. If PETSc cannot solve it (any more?), this case should be excluded from SSA calculations. In fact, I cannot find the difference in code versions?! 
Torsten

**Comment by florianziemen** [comment]
I get a similar crash (4 cores on my local PC)
`PISM ERROR: first  (X) equation in the SSAFD system: zero diagonal entry at a regular (not Dirichlet B.C.) location: i = 101, j = 164`
In this case it is an isolated grid point with an ice thickness of 417.416m (acc. to the crash file).
The local topg is -370.31. Converting the thickness to depth below sea level by *.91/1.028 (ice density, sea water density), I get an ice depth of 369.5m. i.e., this should probably have been calved, but for some reason it was not. Maybe the iceberg remover doesn't consider the sea water density, but the SSA-related code does?

On the positive side, I just completed a 5km 135 kyr Greenland spinup without any problems. 

Cheers

Flo

**Comment by florianziemen** [comment]
Seems to be inconsistent treatment of the sea water density.
Setting it to 999 seems to fix the problem for my pism version (year 5153 and counting without a crash) . @tkleiner, can you please try to reproduce this behavior in standard pism? (change in config.cdl and run make install).

My model version includes a first stab at the problem in SSAFD.cc. Might be helpful. I think, it fixed most problems for me. See attached patch (not meant as permanent solution, but a workaround).
[patch.txt](https://github.com/pism/pism/files/373978/patch.txt)

Cheers

Flo

**Comment by florianziemen** [comment]
Sorry for the confusion and editing in my replies. Seems like I need more coffee... ;)

**Comment by talbrecht** [comment]
Thanks Florian, altering the sea_water_density (999) in the pism_config.cdl solved the problem for me too. But I still don't see the inconsistency in the code: both src/base/stressbalance/ssa/SSAFD.cc and src/base/calving/PISMIcebergRemover.cc consider mask values, that has been calculated before...

**Comment by florianziemen** [comment]
Here’s a couple thoughts and tests before I’ll mostly dive back into other stuff… 

Well, in any case, I think that grid cell should be calved. Maybe we are looking at several problems at the same time.

/src/base/util/Mask.hh L87

`    const bool is_floating = (hfloating > hgrounded + is_floating_thickness),                                                         
`
with` is_floating_thickness = 1.0` (src/config.cdl), this explains why the grid cell was not calved, although the thickness too low to be really grounded. If any other piece of code now does a direct check (or computes a floating-point value based on the ice thickness and the densities), things blow. We can test for that by varying this number (0, 1, 20).

Next question: Any issues with halo/ghosts/… 
In my my run, the whole thing blows at 164/101 which very close to, but not at the cpu boundary (200x200 grid on 4 cores). In Thomas’ 20 core run, the code blows at i = 114, j = 164. That does not look like any multiple of 40 or 50. A different one of my tests blew at 16/136, so definitely not a ghost problem.

Last issue:
The density.
If the lower sea-water density really fixes things (instead of just getting us on a lucky trajectory, Thomas had one crash after 200 years and one after 20,000 ), then the density is handled wrongly somewhere in the code. I don’t see any specific reason, why this whole thing should be surrounded by liquid water and not by liquid mercury. Therefore, we should be able to set the density of sea water to 13690.0, expect weirdo behavior, but a running model (we might start with 1500 or so). It probably is much easier to find a bug that’s occurring b/c of a strong contrast than one that occurs b/c of a factor of 1.028.

— ok 13690 will be very bad for major parts of ANT. Basically anything grounded bsl will become floating and lost… — 

A first try with 1100 got me a different model crash (it assembled, but was not able to solve the SSA system).
1500 yielded a crash for a 2 cm ice cell (again, should be calved, but is protected by the 1m difference threshold)
```
S 00000002.295:   25.32623   13.27770         37.37144 266526.53494
PISM ERROR: ### Rank 1 message:
PISM ERROR: first  (X) equation in the SSAFD system: zero diagonal entry at a regular (not Dirichlet B.C.) location: i = 16, j = 136
```

1500 with 0 tolerance has made it across the first 9000 years w/o crashing. Too soon to call, but much of ANT is destroyed already (i.e. a lot of grid cells went floating and had the chance to cause trouble). I’ll leave it to run for as long as it feels like. 
I guess, with the high sea water density, there simply is too much ice floating and that’s not good for the shelves, but so far, I can’t explicitly exclude any funky effects from is_floating_thickness = 0.0 in the config file, as the 1.0 run crashed before even writing an extra file.

So, my guess for now is that the density just got us on a lucky trajectory, and the real issue is the threshold of 1m, but this is definitely not claiming I found the solution. More tomorrow, when I have more results. Any thoughts on the issue are appreciated.

Cheers

Flo

**Comment by florianziemen** [comment]
Ok, `sea_water_density = 1500` and `is_floating_thickness = 0.0` yields a diverged KSP crash after 33 kyrs. Don't really know what we learn from that. I'll give it one more try with a density of 1028 and 0m tolerance.

```
S 00033046.694:   15.67067    9.37080          2.46378 71538383045433304.00000
PISM WARNING:  KSPSolve() reports 'diverged'; reason = -4 = 'DIVERGED_DTOL'
  writing linear system to PETSc binary file SSAFD_kspdivergederror.petsc ...
  re-trying using the Additive Schwarz preconditioner...

PISM WARNING:  KSPSolve() reports 'diverged'; reason = -4 = 'DIVERGED_DTOL'
  writing linear system to PETSc binary file SSAFD_kspdivergederror.petsc ...

PISM WARNING: output file 'run_30km_stressbalance_failed.nc' already exists. Moving it to 'run_30km_stressbalance_failed.nc~'.

PISM ERROR: stress balance computation failed. Model state was saved to 'run_30km_stressbalance_failed.nc'
```

**Comment by talbrecht** [comment]
Great, I think "is_floating_thickness=1.0" explains the inconsistency in the mask. No failure for me with when setting its value to 0.0 in config! 
Have a nice weekend!
Torsten

**Comment by florianziemen** [comment]
Same here, got 100 000 years of Antarctica spinup w/o a crash. Thomas said, he had some crash-free test runs, too.
@ckhroulev Any thoughts on why there is a 1.0 in the config? Do you see reasons, why we should not set it to 0.0? I assume that whoever added this offset had something in mind...

Cheers, Flo

**Comment by bueler** [comment]
Thomas, Torsten, Flo --

Definitely glad you guys have found a work-around.  By my calendar, Constantine will be back at work around August 1.  This issue is one that he (and all of you) have more understanding of than the other AK people.  (I think; anyway, Andy hasn't piped up.)  Sorry for our silence on this.

Ed

**Comment by ckhroulev** [comment]
Thomas, Torsten, Flo,

The configuration parameter `mask_is_floating_thickness_standard` is there for "historical reasons," which is just a way of saying "for no good reason at all, as far as I can tell." Setting it to zero or removing it altogether sounds OK to me. It would definitely get rid of a possible inconsistency in the way the floatation criterion is applied.

Regarding crashes with messages such as "`PISM ERROR: first (X) equation in the SSAFD system: zero diagonal entry at a regular (not Dirichlet B.C.) location: i = XX, j = YY`": I made PISM stop with this error when the PETSc solver would fail and give a less helpful error message ("KSP diverged" or "zero pivot in LU factorization"). My goal was to provide a bit more information (the location of the problematic grid point).

Regarding crashes with `-subgl`: note that setting `mask_is_floating_thickness_standard` to zero may be enough in most cases, but is not sufficient: the 1D (flow-line) parameterization of the grounding line position is hard to extend to 2D, and strictly speaking [this](https://github.com/pism/pism/blob/stable0.7/src/base/grounded_cell_fraction.cc#L224) is wrong.

In particular, when `lambda_x` is zero, the resulting 2D grounded cell fraction will be zero even if `lambda_y` is not zero:

``` c++
result(i, j) = lambda_x * lambda_y;
```

In the `dev` branch I added the option (and the config. flag) `ssafd_replace_zero_diagonal_entries` (enabled by default). When it is set, PISM sets the basal drag term ("beta") to a large value (`beta_ice_free_bedrock`, 1.8e9 Pa s / m by default) at grid points that would otherwise produce a zero diagonal entry. This "disables" sliding in problematic areas and eliminates at least some of these crashes.
